bounded-collections/src/bounded_btree_set.rs

```diff
@@ -29,9 +34,67 @@ use core::{borrow::Borrow, marker::PhantomData, ops::Deref};
 ///
 /// Unlike a standard `BTreeSet`, there is an enforced upper limit to the number of items in the
 /// set. All internal operations ensure this bound is respected.
+#[cfg_attr(feature = "serde", derive(Serialize), serde(transparent))]
 #[derive(Encode, scale_info::TypeInfo)]
 #[scale_info(skip_type_params(S))]
-pub struct BoundedBTreeSet<T, S>(BTreeSet<T>, PhantomData<S>);
+pub struct BoundedBTreeSet<T, S>(BTreeSet<T>, #[cfg_attr(feature = "serde", serde(skip_serializing))] PhantomData<S>);
+
+#[cfg(feature = "serde")]
+impl<'de, T, S: Get<u32>> Deserialize<'de> for BoundedBTreeSet<T, S>
+where
+	T: Ord + Deserialize<'de>,
+	S: Clone,
+{
+	fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+	where
+		D: Deserializer<'de>,
+	{
+		// Create a visitor to visit each element in the sequence
+		struct BTreeSetVisitor<T, S>(std::marker::PhantomData<(T, S)>);
+
+		impl<'de, T, S> Visitor<'de> for BTreeSetVisitor<T, S>
+		where
+			T: Ord + Deserialize<'de>,
+			S: Get<u32> + Clone,
+		{
+			type Value = BTreeSet<T>;
+
+			fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+				formatter.write_str("a sequence")
+			}
+
+			fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+			where
+				A: SeqAccess<'de>,
+			{
+				let size = seq.size_hint().unwrap_or(0);
+				let max = match usize::try_from(S::get()) {
+					Ok(n) => n,
+					Err(_) => return Err(A::Error::custom("can't convert to usize")),
+				};
+				if size > max {
+					Err(A::Error::custom("out of bounds"))
+				} else {
+					let mut values = BTreeSet::new();
+
+					while let Some(value) = seq.next_element()? {
+						values.insert(value);
+						if values.len() > max {
```

Comment:
This is weird. You insert before checking for the max, i.e. you add 1 more value over the limit before you check the limit?

Reply:
I also found it weird. It was like that in `BoundedVec`'s Deserialize implementation: https://github.com/paritytech/parity-common/blob/master/bounded-collections/src/bounded_vec.rs#L90-L92

I thought that was intentional and didn't put much thought tbh. 

I can fix them both if you like?

Reply:
Sure, but there's one more additional issue: checking `values.len() > max` doesn't sound right to me either, because it too is also waiting for 1 element over the max before the condition triggers. What should really happen is values.len() >= max`.

It would be so great if you can write tests for these as well.

Reply:
Since we are already discussing the implementation of Deserialize for BoundedVec as well, I had another thing which confused me:

https://github.com/paritytech/parity-common/blob/master/bounded-collections/src/bounded_vec.rs#L84-L85

So, if we are performing a check in here for the size, do we need another check in here: 
https://github.com/paritytech/parity-common/blob/master/bounded-collections/src/bounded_vec.rs#L90-L92
?

I think the second check is not necessary and the error produced by the second check is unreachable. So, I don't know if it's possible to write a test for triggering that error. It will trigger the previous error instead.

I might be missing something, so feel free to correct me

Reply:
The first check is using `size_hint()` and that could also be `None` in which case the size is `0` as you see in the code.

Reply:
I've added a new commit:
- insertion is now performed after the bound check
- new tests for bound check:
    - i've copied the `deserializer_test` and made the size equal to the bound
    - for the tests, `size_hint()` does indeed return `None`, so I'm positive I'm testing the inner check
    - if you prefer not to have an additional test, the same test can be done by changing the bound to the size of the deserialized string in the `deserializer_test`